### DISCUSSION TITLE
How to communicate with the backend

### DISCUSSION THREAD

**Discussion by wassimkha** [eclipse-theia/theia#9537] [Q&A] [answered]
I'm trying to communicate with the backend, receiving a simple string from the backend, but when calling the backend, the code stops and nothing happens. Could someone help?

What I have so far:

1 service (my-service-protocol):

```ts
import {injectable} from "inversify";

export const MyServicePath = '/services/my-service';


export const MyService = Symbol('MyService');
export interface MyService {
    getNameString(name: string): Promise<string>
}

@injectable()
export class MyServiceImpl implements MyService {
  getNameString(name: string): Promise<string> {
    return new Promise<string>(resolve => resolve('Hello ' + name));
  }
}
```

my-service-module:
```ts
import { ContainerModule } from 'inversify';
import { ConnectionHandler, JsonRpcConnectionHandler } from '@theia/core/lib/common';
import {MyService, MyServicePath, MyServiceImpl} from "../common/my-service-protocol";

export default new ContainerModule(bind => {
  bind(MyServiceImpl).toSelf().inSingletonScope();
  bind(MyService).toService(MyServiceImpl);
  bind(ConnectionHandler).toDynamicValue(context => new JsonRpcConnectionHandler(MyServicePath, () => context.container.get(MyService))).inSingletonScope();
});
```

extention contribution:

```ts
export const MyServiceExtensionCommand = {
  id: 'HelloWorldExtension.command',
  label: "Shows a message"
};
@injectable()
export class MyServiceCommandContibution implements CommandContribution {
  constructor(
    @inject(MessageService) private readonly messageService: MessageService,
    @inject(MyService) protected readonly myService: MyService
  ) { }

  registerCommands(registry: CommandRegistry): void {
    registry.registerCommand(MyServiceExtensionCommand, {
      execute: async () => {
        //const env = await this.myService.getEnvVariables();
        //this.messageService.info('Environment variables from the server: ' + JSON.stringify(env));
        const setting = await this.myService.getNameString("name");
        await this.messageService.info(setting);
      }
    });
  }
}
```
the extension module (how I'm binding it). The last two binds relate to the backend
```ts
export default new ContainerModule(bind => {
    // register new widget (literally sidebar button)
    bindViewContribution(bind, ExtensionTestContribution);
    bind(FrontendApplicationContribution).toService(ExtensionTestContribution);
    bind(ExtensionTestWidget).toSelf();
    bind(WidgetFactory).toDynamicValue(ctx => ({
        id: ExtensionTestWidget.ID,
        createWidget: () => ctx.container.get<ExtensionTestWidget>(ExtensionTestWidget)
    })).inSingletonScope();

    bind(CommandContribution).to(MyServiceCommandContibution).inSingletonScope();
    bind(MyService).toDynamicValue(context => WebSocketConnectionProvider.createProxy(context.container, MyServicePath)).inSingletonScope();

 
});
```
Then I try to call it from my widget like this:
```ts
@inject(MyService)
    protected readonly myService: MyService;
```
```ts
const app_name = await this.myService.getNameString("wassim");
console.log(app_name)
```
once it reach ```const app_name = await this.myService.getNameString("wassim"); ``` the code stops, it's not getting data from the backend. What is missing?

Any help would be greatly appreciated!

**Comment by vince-fugnitto** [comment]
@wassimkha thank you for the discussion :+1: 

I haven't looked at the code thoroughly yet but I thought I'd mention that we have a `backend communication` template as part of our generator which can certainly inspire you for your own implementation as it sets in place the necessary parts to be able to communicate between the `frontend` and `backend`.

- https://github.com/eclipse-theia/generator-theia-extension
- https://github.com/eclipse-theia/generator-theia-extension/blob/master/templates/backend/README.md

**Reply by wassimkha** [reply]
Hi Vince, Yeah I've been trying to follow that template without much luck :/

**Reply by wassimkha** [reply]
@vince-fugnitto 
I retried following the complete backend template, i keep getting the following error:
```bash 
ERROR Cannot find a service for the path: /services/helloBackend
```
When I try to use the service as such (from my react widget):
```ts
@injectable()
export class ExtensionTestWidget extends ReactWidget {

    @inject(HelloBackendService)
    protected readonly backendService: HelloBackendService;
...(other code)
this.backendService.sayHelloTo("wassim")
          .then(a => console.log(a))
          .catch(e => console.log(e))
}
```
Might It have something to do with the fact that I'm not injecting it correctly in the widget?

**Reply by vince-fugnitto** [reply]
@wassimkha if you don't mind pushing the code to a public repository I can help take a look at it with you. It'd be easier to follow and identify what might be missing :)

**Reply by wassimkha** [reply]
@vince-fugnitto sounds good, thank you for your time :)

[Here's the code](https://github.com/wassimkha/theia-test-extension)

**Reply by vince-fugnitto** [reply]
> @vince-fugnitto sounds good, thank you for your time :)
> 
> [Here's the code](https://github.com/wassimkha/theia-test-extension)

The first thing I noticed is that the separation for the extension is not consistent with the template, `browser`, `common` and `node` separation. A cleanup is likely necessary first:

_Extension_:

![image](https://user-images.githubusercontent.com/40359487/120337947-c4227000-c2c1-11eb-8882-957a566aea2e.png)

_Template_:

![image](https://user-images.githubusercontent.com/40359487/120338088-e1efd500-c2c1-11eb-91b2-fdbc95ae1a51.png)

We need to remember that `browser` is the `frontend` implementation (cannot make use of `node` apis), `common` is common to both `frontend` and `backend`, and `backend` contains the `node` api without `frontend` code:

More Info: https://github.com/eclipse-theia/theia/wiki/Code-Organization.

**Reply by wassimkha** [reply]
You're right, I cleaned the code and re-pushed

**Reply by wassimkha** [reply]
After a while, still stuck on the error
```
ERROR Cannot find a service for the path: /services/helloBackend
```

Any clues?

**Reply by vince-fugnitto** [reply]
@wassimkha you are missing the `backend` declaration of your extension's `package.json` to declare the backend bindings:

```json
"theiaExtensions": [
    {
      "frontend": "lib/browser/extension-test-frontend-module",
      "backend": "lib/node/my-service.module"
    }
  ]
  ```

Location: 
- https://github.com/wassimkha/theia-test-extension/blob/324e2eb70d61942d17478423745b00f671a7fa13/extension-test/package.json#L27.

Output:

```
$ theia start
root INFO Theia app listening on http://localhost:3000.
root INFO Configuration directory URI: 'file:///home/evinfug/.theia'
root INFO Changed application state from 'ready' to 'closing_window'.
root INFO >>> Storing the layout...
root INFO <<< The layout has been successfully stored.
root INFO >>> Stopping frontend contributions...
root WARN Frontend CommonFrontendContribution.configure is slow, took: 346.1 ms
root INFO Detected keyboard layout from browser API: US (PC)
root INFO Changed application state from 'init' to 'started_contributions'.
root INFO Changed application state from 'started_contributions' to 'attached_shell'.
root INFO >>> Restoring the layout state...
root INFO <<< The layout has been successfully restored.
root INFO Changed application state from 'attached_shell' to 'initialized_layout'.
root INFO Changed application state from 'initialized_layout' to 'ready'.
root INFO Hello World
root INFO Changed application state from 'ready' to 'closing_window'.
root INFO >>> Storing the layout...
root INFO <<< The layout has been successfully stored.
root INFO >>> Stopping frontend contributions...
root INFO <<< All frontend contributions have been stopped.
```

namely (after executing the command):

```
root INFO Hello World
```

:smile:

**Reply by wassimkha** [reply]
@vince-fugnitto you're a legend!! It's working now, thanks!

**Comment by vince-fugnitto** [comment, marked as answer]
Fixed thanks to: https://github.com/eclipse-theia/theia/discussions/9537#discussioncomment-811917.
